Add changelog (#26894)
solana-labs/solana: CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and follows a [Backwards Compatability Policy](https://docs.solana.com/developing/backwards-compatibility)

## Adding to this Changelog
### Audience
* Entries in this log are intended to be easily understood by consensus validators, rpc operators, and dapp developers.

### Noteworthy
* Changes that relate to a SIMD, or add a feature gate should add an entry to this log. Other changes that are relevant to the aforementioned audiences should add an entry at the discretion of authors and reviewers.

### Instructions

> * Update this log with the same commit that implements the change. If the change is spread over several commits update this log with the commit that makes the feature code complete.

Author: I've updated the wording to remove ambiguity, but I think we're envisioning the same thing here. I agree that asking authors to update the PR once they know the PR number is too much friction.

* Add notes to the [Unreleased] section in each branch that you merge to.
  * Add a description of your change to the Changes section
  * If your change is likely to require operators to update their configs add to the Upgrade Notes
* If you add entries on multiple branches please use the same wording if possible. This simplifies the process of diffing between versions of the log. 

> ## How To Maintain This Changelog

Reviewer: I was thinking we could first try really only adding to the CHANGELOG in `master` as new features are developed. When a new release branch is created, current contents of CHANGELOG go into the 1.x.0 release nodes and then the file is reset.

We then follow our current manual process for 1.x.y patch release notes for now (maybe consider doing something more later, but baby steps!)

This also reenforces the model that features are not backported (!!), only bug fixes

Author: Ok that's reasonable. I've removed the line relating to backports.

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
and follows a [Backwards Compatability Policy]
(https://docs.solana.com/developing/backwards-compatibility)

## Adding to this Changelog
### Audience
* Entries in this log are intended to be easily understood by consensus
validators, rpc operators, and dapp developers.

### Noteworthy
* Changes that relate to a SIMD, or add a feature gate should add an entry to
this log. Other changes that are relevant to the aforementioned audiences should
add an entry at the discretion of authors and reviewers.

### Instructions
* Update this log in the same pull request that implements the change. If the
change is spread over several pull requests update this log in the one that
makes the feature code complete.

> * Add notes to the [Unreleased] section in each branch that you merge to.

Reviewer 2: Is the `Unreleased` section features that would not get immediately released in each version?

Reviewer 2: So this would be something like RuntimeV2 which has a ton a subtasks but would live in `Unreleased` until all subtasks are complete?

Author: https://keepachangelog.com/en/1.0.0/  envisions [Unreleased] to just be the notes for the next version. If we adopt that practice then I like @mvines suggestion of just numbering it when we bump the version number.

I don't think we need a section for partially complete work, since we already have guidance to add a log entry when the work is code complete.

  * Add a description of your change to the Changes section
  * If your change is likely to require operators to update their configs add to

>   the Upgrade Notes

Author: In this comment is "clients" referring to things like our CLI, or things like solana-validator?

>   the Upgrade Notes

Reviewer 3: cli, wallets, explorers, etc

> ## How To Maintain This Changelog

Reviewer 3: gonna wait 'til current comments are resolved to further review this section 'cause github doesn't support hiding only comments from other reviewers

* When a release is tagged and the version is bumped:

>   * Date and tag the [Unreleased] section with the new release including a link
>   to the release page

Reviewer: ```suggestion
  * Use the current contents of the [Unreleased] section to populate the release notes
```

Reviewer 3: neither of these is my understanding of the intended use of `Unreleased`.  the way i read it is that `Unreleased` is for any ongoing changes that are not yet complete. it lives independently of any version and items that cannot be completed in one change set live here during their development, before being moved into the appropriate release upon completion. that is `Unreleased` is _not_ versioned or dated in anyway. it's just a "we're actively working on this stuff, you'll get it later"

>   * Create new [Unreleased] section

Reviewer: ```suggestion
  * Clear the contents of the [Unreleased] section
```

change

> ## Unreleased

Reviewer: bikeshed:
```suggestion
## New Features
```

* Changes
  * Added a changelog!
* Upgrade Notes

## Released

> ### 2023-05-31 - [1.16.0](https://github.com/solana-labs/solana/releases/tag/v1.16.0)

Reviewer: Over time the release section would only contain the x.y.0 release right? We don't add a new entry for patch releases? How about DRYing this out and just linking generically to https://github.com/solana-labs/solana/releases

Author: > All notable changes to this project will be documented in this file.

I think this should be our guiding principle. If we have the changelog in each branch, and use it to document "all notable changes" it reduces mental load on both authors and consumers. The only question of where to find information is which branch to look at, but anyone reading the changelog likely already knows our branching stradegy.

I'm envisioning:
* The log maintains a full history since its inception (newest content at the top)
* Both bug fixes and new features are included (if they're notable)
* Patch releases get a new section (maybe skip the new section if there's no new content?)
* Henceforth the x.y.0 release should be the mergebase of master and vx.y

So the master version of CHANGELOG.md will include only the x.y.0 releases, but then the release branches will each contain a few more entries for bug fixes that were backported. That way the changelog mimics the code changes and the branching strategy is familiar and consistent.

Reviewer: Ah, ok. Got it and I think I like it. One matter though: we bump crate versions *after* tagging a release, but with changelog would be introducing a step *before* tagging a release to update the changelog with the new release. We're 100% going to screw this up, a lot. 

Instead how about we bump the changelog at the same time as the crate versions? There's never an "Unreleased" section at all then, the most recent section (like the latest crate version of a branch) is the next anticipated release version

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
and follows a [Backwards Compatability Policy]
(https://docs.solana.com/developing/backwards-compatibility)

> ## Unreleased

Reviewer: The "Unreleased" and "Released" section headers are still somewhat problematic for tagging, as they'll require a commit BEFORE tagging. 

I was thinking we'd just have a single list, with 1.17.0 at the top in this case, where new entries are added. `./scripts/increment-cargo-version.sh` then adds a new entry above it, be it 1.17.1 or 1.18.0, as it runs to update the rest of the version numbers post-tagging

Reviewer 3: it'd be nice if we can denote the "working" release somehow. as described, that information would need recovered from another source. i'm struggling to see how we do that without another commit (and obnoxious additional release coordination) though

Author: I'll have a go at a script to generate the releases. My preference is to avoid BuildKite but open to that if anyone sees benefit of doing it there.

Reviewer: Cool. A buildkite job with an [input step](https://buildkite.com/docs/pipelines/input-step) to specify the kind of release seems like it could be pretty convenient: no need to check out the exact tree to ship locally, or dork around with a github access token

Reviewer 3: we decided (last week) that this script would be the way forward, but ought not block this pr.  resolving this convo accordingly

### [1.17.0]
* Changes
  * Added a changelog!
* Upgrade Notes

## Released
### 2023-05-31 - [1.16.0](https://github.com/solana-labs/solana/releases/tag/v1.16.0)

## Adding to this Changelog
### Audience
* Entries in this log are intended to be easily understood by core contributors,
consensus validator operators, rpc operators, and dapp developers.

### Noteworthy
* A change is noteworthy if it:
  * Adds a feature gate, or
  * Implements a SIMD, or

>   * Modifies API or ABI, or

Reviewer: ```suggestion
  * Modifies a public API, or
```
I removed ABI since that seems to be already covered by "Implements a SIMD", but I don't feel too strong

  * Changes normal validator / rpc run configurations, or
  * Changes command line arguments, or
  * Fixes a bug that has received public attention, or
  * Significantly improves performance, or

>   * Is authored by an external contributor.

Reviewer 3: i was hoping we could filter the refspec compare query by author, like the branch commits list can be. then we could just bash up a list of links for each external contributor's commits in the version rather than call out each individual contribution. doesn't look like that's possible today though. hopefully this doesn't turn into a dumb scoreboard.

Reviewer 4: Agreed that calling out each individual contribution could be a lot.

Throwing out an idea I saw elsewhere - [rust-rocksdb has a section specifically for first time contributors](https://github.com/rust-rocksdb/rust-rocksdb/releases). Could potentially do something similar where first time external contributors always get a shout out, and then subsequent changes are subject to the above criteria.

I'm not sure how that list is generated tho, and `rust-rocksdb` has a much lower commit volume than this repo so it is possible that the list is created manually. Not sure we'd want to take that on

### Instructions
* Update this log in the same pull request that implements the change. If the
change is spread over several pull requests update this log in the one that
makes the feature code complete.
* Add notes to the [Unreleased] section in each branch that you merge to.
  * Add a description of your change to the Changes section.
  * Add Upgrade Notes if the change is likely to require:
    * validator or rpc operators to update their configs, or
    * dapp developers to make changes.
* Link to any relevant feature gate issues or SIMDs.
* If you add entries on multiple branches use the same wording if possible.
This simplifies the process of diffing between versions of the log.

## Maintaining This Changelog

> * When a release is tagged and the version is bumped:

Reviewer 3: this is referring to a _patch_ release, right?

how do we want to manage master given that we never actually tag a release off it? seems like we'll just have a bunch of .0's, some items from each being duplicated in beta/stable branches when backported

Author: I'm thinking henceforth we always tag the x.y.0 release immediately when a branch is created. Generally the x.y.0 release will be the mergebase of vx.y and master (barring shenanigans like v1.11 becoming v1.14)

Reviewer 3: right, but we never tag releases off master, so the pedantic interpretation of the process here implies that the master changelog will be a never ending `Unreleased` section

Author: How about this (done atomically):
* Commit to master updating the changelog: 
    * Remove `Unreleased` annotation from vx.y.0 section.
    * Create new section: `vx.y+1.0 - Unreleased`
* Create vx.y branch starting at that commit
* Tag that commit as vx.y.0

Reviewer 3: seem's like a good enough place to start :shipit:

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

> See [Adding to this Changelog](#adding-to-this-changelog) section below for details on how to add notes to this file

Reviewer 3: it might make sense to cite our [backward compatibility policy](https://docs.solana.com/developing/backwards-compatibility) in this paragraph as well

(https://docs.solana.com/developing/backwards-compatibility)

## Unreleased
### [1.17.0]
* Changes
  * Added a changelog!
* Upgrade Notes

## Released
### 2023-05-31 - [1.16.0](https://github.com/solana-labs/solana/releases/tag/v1.16.0)

## Adding to this Changelog
### Audience
* Entries in this log are intended to be easily understood by contributors,
consensus validator operators, rpc operators, and dapp developers.

### Noteworthy
* A change is noteworthy if it:
  * Adds a feature gate, or
  * Implements a SIMD, or
  * Modifies a public API, or
  * Changes normal validator / rpc run configurations, or
  * Changes command line arguments, or
  * Fixes a bug that has received public attention, or
  * Significantly improves performance, or
  * Is authored by an external contributor.

### Instructions
* Update this log in the same pull request that implements the change. If the
change is spread over several pull requests update this log in the one that
makes the feature code complete.
* Add notes to the [Unreleased] section in each branch that you merge to.
  * Add a description of your change to the Changes section.
  * Add Upgrade Notes if the change is likely to require:

>     * validator or rpc operators to update their configs, or

Reviewer 5: Would be good to develop a standard to highlight this to those that are affected. Could be a marker
* [\<insert affected list here\>] Explanation of the change and how it affects each group

Author: I'm inclined to leave this sort of annotation as a step for comms, but I'm open to asking engineers to do it if anyone feels strongly.

    * dapp developers to make changes.
* Link to any relevant feature gate issues or SIMDs.
* If you add entries on multiple branches use the same wording if possible.
This simplifies the process of diffing between versions of the log.

> ## Maintaining This Changelog

Reviewer 5: How do we feel about other automated tools to help [visualize the changes](https://livablesoftware.com/tools-to-visualize-the-history-of-a-git-repository/) ? I don't have strong feelings but considering the diverse set of viewers it would be good to add some automated tools that makes it more accessible where possible.

Reviewer: Seems out of scope for this effort

# Changelog
All notable changes to this project will be documented in this file.

Please follow the [guidance](#adding-to-this-changelog) at the bottom of this file when making changes
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
and follows a [Backwards Compatability Policy]
(https://docs.solana.com/developing/backwards-compatibility)

> Release branches have their own copy of this changelog:
> * [beta - v1.16](https://github.com/solana-labs/solana/blob/v1.16/CHANGELOG.md)
> * [stable - v1.14](https://github.com/solana-labs/solana/blob/v1.14/CHANGELOG.md)
> 
> ## [1.17.0] - Unreleased

Reviewer 3: ```suggestion
Release channels have their own copy of this changelog:
* [edge - v1.17](#edge-channel)
* [beta - v1.16](https://github.com/solana-labs/solana/blob/v1.16/CHANGELOG.md)
* [stable - v1.14](https://github.com/solana-labs/solana/blob/v1.14/CHANGELOG.md)

<a name="edge-channel"></a>
## [1.17.0] - Unreleased
```

* switch language from "branches" to "channels"
* adds entry for edge channel
